src/orderbook.ts

```diff
@@ -0,0 +1,79 @@
+export class Price {
+  numerator: number;
+  denominator: number;
+
+  constructor(numerator: number, denominator: number) {
+    this.numerator = numerator;
+    this.denominator = denominator;
+  }
+
+  toNumber() {
+    return this.numerator / this.denominator;
+  }
+
+  toJSON() {
+    return this.toNumber();
+  }
+}
+
+export class Offer {
+  price: Price;
+  volume: number;
+
+  constructor(price: Price, volume: number) {
+    this.volume = volume;
+    this.price = price;
+  }
+
+  toJSON() {
+    return {price: this.price, volume: this.volume};
+  }
+}
+
+export class Orderbook {
+  baseToken: string;
+  quoteToken: string;
+  private asks: Map<number, Offer>;
+  private bids: Map<number, Offer>;
+
+  constructor(baseToken: string, quoteToken: string) {
+    this.baseToken = baseToken;
+    this.quoteToken = quoteToken;
+    this.asks = new Map();
+    this.bids = new Map();
+  }
+
+  addBid(bid: Offer) {
+    addOffer(bid, this.bids);
+  }
+
+  addAsk(ask: Offer) {
+    addOffer(ask, this.asks);
+  }
+
+  toJSON() {
```

Comment:
nit: Why the name is `toJSON`. I'd suggest to give a functional name and type to the object containing bids and asks, for example BookOffers and call the function `getBookOffers`

Reply:
toJSON is the function that JSON.stringify calls on the objects it's trying to serialize. Thus we need to implement it in order to ensure the serialization happens in the way we want to. I can create a getBookOffers method on top of that, but it wasn't needed until now.

Reply:
fair enough